lib/active_merchant/billing/gateways/globalcollect.rb

```diff
@@ -0,0 +1,294 @@
+            "countryCode" => shipping_address[:country]
+          }
+          post["order"]["customer"]["shippingAddress"]["name"] = {
+            "firstName" => shipping_address[:firstname],
+            "surname" => shipping_address[:lastname]
+          }
+        end
+      end
+
+      def parse(body)
+        JSON.parse(body)
+      end
+
+      def url(action, authorization)
+        (test? ? test_url : live_url) + uri(action, authorization)
+      end
+
+      def uri(action, authorization)
+        uri = "/v1/#{@options[:merchant_id]}/"
+        case action
+        when :test_connection
+          uri + "services/testconnection"
+        when :authorize
+          uri + "payments"
+        when :capture
+          uri + "payments/#{authorization}/approve"
+        when :refund
+          uri + "payments/#{authorization}/refund"
+        when :void
+          uri + "payments/#{authorization}/cancel"
+        end
+      end
+
+      def commit(action, post, authorization = nil)
+        begin
+        response = parse(ssl_post(url(action, authorization), post.to_json, headers(action, post, authorization)))
+        rescue ResponseError => e
+          raise unless(e.response.code.to_s =~ /4\d\d/)
+          response = parse(e.response.body)
+        end
+
+        succeeded = success_from(response)
+        Response.new(
+        succeeded,
+        message_from(succeeded, response),
+        response,
+        authorization: authorization_from(succeeded, response),
+        error_code: error_code_from(succeeded, response),
+        test: test?
+        )
+
+      end
+
+      def headers(action, post, authorization = nil)
+        {
+          "Content-type"  => content_type,
+          "Authorization" => auth_digest(action, post, authorization),
+          "Date" => date
+        }
+      end
+
+      def auth_digest(action, post, authorization = nil)
+        data = <<-EOS
+POST
+#{content_type}
+#{date}
+#{uri(action, authorization)}
+EOS
+        digest = OpenSSL::Digest.new('sha256')
+        key = @options[:secretApiKey]
+        "GCS v1HMAC:#{@options[:apiKeyId]}:#{Base64.strict_encode64(OpenSSL::HMAC.digest(digest, key, data))}"
+      end
+
+      def date
+        @date ||= Time.now.strftime("%a, %d %b %Y %H:%M:%S") + " GMT" # Must be same in digest and HTTP header
+      end
+
+      def content_type
+        "application/json"
+      end
+
+      def success_from(response)
+        !response["errorId"]
+      end
+
+      def message_from(succeeded, response)
+        if succeeded
+          "Succeeded"
+        else
+          response["errors"][0]["message"] || "Unable to read error message"
+        end
+      end
+
+      def authorization_from(succeeded, response)
+        if succeeded
+          response["id"] || response["payment"]["id"] || response["paymentResult"]["payment"]["id"]
+        else
+          response["errorId"]
+        end
+      end
+
+      def error_code_from(succeeded, response)
```

Comment:
I wonder if `response[:errorId]` isn't the better error code to use (it's what's used to determine if it was an unsuccessful response)?